lib/src/trie/minimize_proof.rs

```diff
@@ -0,0 +1,82 @@
+use hashbrown::HashSet;
+use itertools::Itertools;
+
+use crate::trie::{
+    bytes_to_nibbles,
+    proof_decode::{DecodedTrieProof, StorageValue},
+    proof_encode::ProofBuilder,
+};
+use alloc::vec::Vec;
+
+pub enum MinimizeProofError {
+    KeyDoesNotMatch,
+    IncompleteProof,
+}
+
+/// Minimizes a single-key proof, removing all entries that are not related to
+/// the proof for that key.
+///
+/// Returns the resulting proof encoded
+pub fn minimize_proof<T: AsRef<[u8]>>(
+    decoded_proof: &DecodedTrieProof<T>,
+    trie_root_merkle_value: &[u8; 32],
+    key: &[u8],
+) -> Result<Vec<u8>, MinimizeProofError> {
+    let mut builder = ProofBuilder::new();
+
+    let nibbles = decoded_proof
+        .closest_ancestor_in_proof(
+            trie_root_merkle_value,
+            bytes_to_nibbles(key.iter().copied()),
+        )
+        .map_err(|_| MinimizeProofError::IncompleteProof)?
+        .ok_or(MinimizeProofError::KeyDoesNotMatch)?
+        .collect_vec();
+
+    // Set the node value of the leaf
+    let node = decoded_proof
+        .trie_node_info(trie_root_merkle_value, nibbles.iter().cloned())
+        .map_err(|_| MinimizeProofError::IncompleteProof)?;
+    let storage_value = match node.storage_value {
+        StorageValue::Known { value, .. } => Some(value),
+        _ => None,
+    };
+    builder.set_node_value(&nibbles, node.node_value, storage_value);
+
+    // Query a missing node and provide its value. Stop when the proof is complete.
+    let mut maybe_missing = builder.missing_node_values().next().map(|v| Vec::from(v));
+    while let Some(ref missing) = maybe_missing {
+        let value = decoded_proof
+            .trie_node_info(trie_root_merkle_value, missing.iter().copied())
+            .map_err(|_| MinimizeProofError::IncompleteProof)?
+            .node_value;
+        builder.set_node_value(&missing, value, None);
+        maybe_missing = builder.missing_node_values().next().map(|v| Vec::from(v));
+    }
```

Comment:
```suggestion
    loop {
        let Some(maybe_missing) = builder.missing_node_values().next().map(|v| Vec::from(v)) else { break; };
        let value = decoded_proof
            .trie_node_info(trie_root_merkle_value, missing.iter().copied())
            .map_err(|_| MinimizeProofError::IncompleteProof)?
            .node_value;
        builder.set_node_value(&missing, value, None);
    }
```

Seems easier to understand this way IMO.